データベース設計

Yokko405 opened the pull request:
What
データベース設計

Why
フリーマーケットサイトを再現するために、
ER図の作成とREADMEへデータベース設計を記述しました。
対象機能は以下3点です。
・ユーザー管理機能
・商品出品機能
・商品購入機能

"""
▼ER図
https://gyazo.com/e67998a36c49ee3bf47c85330bc3c7eb

Yokko405 committed:
データベース設計

TakashiKozuma commented on README.md:
```diff
@@ -1,24 +1,47 @@
+|----------------|--------|-------------|-----------|
+| nickname       | string | null: false | ニックネーム |
+| email          | string | null: false | メールアドレス |
+| password_digest| string | null: false | パスワード   |
```
```Rb
| password_digest| string | null: false | パスワード |
```
こちらのカラムはdeviseのgemを使用しての実装になるため、deviseが予め用意している
encrypted_passwordというカラムを用意してあげましょう。
passwordを直接保存するのはセキュリティ上危険であるため、暗号化した上でencrypted_passwordというカラムに保存するようになっています。

TakashiKozuma commented on README.md:
```diff
@@ -1,24 +1,47 @@
+| Column         | Type   | Options     | memo      |
+|----------------|--------|-------------|-----------|
+| nickname       | string | null: false | ニックネーム |
+| email          | string | null: false | メールアドレス |
```
```rb
| email | string | null: false | メールアドレス |
```
下記のユーザー管理機能の実装条件に合うように、Optionsに記述を追記しましょう。
- メールアドレスが一意性であること

TakashiKozuma commented on README.md:
```diff
@@ -1,24 +1,47 @@
+
+| Column              | Type       | Options                        | memo                                 |
+|---------------------|------------|--------------------------------|--------------------------------------|
+| user_id             | references | null: false, foreign_key: true | 出品者のユーザーID、usersテーブルの外部キー |
```
```Rb
| user_id | references | null: false, foreign_key: true | 出品者のユーザーID、usersテーブルの外部キー |
```
references 型で記述する際は、カラム名の_idは不要なので修正しましょう。

TakashiKozuma commented on README.md:
```diff
@@ -1,24 +1,47 @@
+| Column              | Type       | Options                        | memo                                 |
+|---------------------|------------|--------------------------------|--------------------------------------|
+| user_id             | references | null: false, foreign_key: true | 出品者のユーザーID、usersテーブルの外部キー |
+| image               | image      | null: false                    | 商品画像                             |
```
```Rb
| image | image | null: false | 商品画像 |
```
アクティブストレージを用いて実装する場合、chat-appのように画像のカラムは不要となるので、このカラムは削除しましょう。

TakashiKozuma commented on README.md:
```diff
@@ -1,24 +1,47 @@
+| condition           | string     | null: false                    | 商品の状態                           |
+| shipping_fee_payer  | string     | null: false                    | 配送料の負担                         |
+| shipping_origin     | string     | null: false                    | 配送元の地域                         |
+| shipping_days       | integer    | null: false                    | 発送までの日数                       |
```
```rb
| category            | string     | null: false                    | カテゴリー                           |
| condition           | string     | null: false                    | 商品の状態                           |
| shipping_fee_payer  | string     | null: false                    | 配送料の負担                         |
| shipping_origin     | string     | null: false                    | 配送元の地域                         |
| shipping_days       | integer    | null: false                    | 発送までの日数                       |
```
アクティブハッシュを用いたカラム名は末尾に_idを付け、stringではなくintegerで実装しましょう。
後の実装で、アソシエーションを用いてデータを取り出すときに必要になってきます。
また、_idの前の単語はモデルの名前になるので単数形で記述しましょう。

TakashiKozuma commented on README.md:
```diff
@@ -1,24 +1,47 @@
+
+## orders テーブル
+
+| Column             | Type       | Options                        | memo                                 |
```
「誰が」「何を」購入したかという購入履歴のテーブルと、配送先の住所のテーブルは分けて実装しましょう。
それぞれの情報の性質が異なるためです。

TakashiKozuma reviewed (commented):
レビュー依頼ありがとうございます！
確認させていただきました。
コメントを致しましたので、ご確認お願いします。
また、次回レビュー依頼時に最新状態のER図の添付もお願いいたします。

修正後はコミット・プッシュのし忘れがないかご確認の上、再度フォームからレビュー依頼をお願い致します。

Yokko405 committed:
データベース設計PR修正

password_digestをencrypted_passwordに変更。
emailカラムに一意性制約（unique: true）を追加。
references型のカラム名から_idを削除（user_idをuserに変更）。
imageカラムを削除（アクティブストレージを使用するため）。
アクティブハッシュを使用するカラムの型をstringからintegerに変更し、カラム名の末尾に_idを追加し、単数形に変更（例：categoryをcategory_idに変更）。
purchase_recordsテーブルとshipping_addressesテーブルを分けて実装。

Yokko405 commented:
ご指摘いただきました6点について、修正が終わりました。ご確認をお願いします。

▼ER図
https://gyazo.com/1e8e1d1e8fed598667c7b5cb7bafb677

minachanchan commented on README.md:
```diff
@@ -1,24 +1,52 @@
-* Deployment instructions
-
-* ...
+# テーブル設計
```
現在アソシエーションについての記載がございません。
**各テーブル同士のアソシエーションについてもご思考いただき、追記をお願いいたします！**

minachanchan commented on README.md:
```diff
@@ -1,24 +1,52 @@
+| payment_amount | integer    | null: false                    | 支払金額 |
+| card_number    | integer    | null: false                    | カード情報 |
+| expiration_date| integer    | null: false                    | 有効期限 |
+| security_code  | integer    | null: false                    | セキュリティコード |
```
```ruby
| order_date     | date       | null: false                    | 注文日 |
| payment_amount | integer    | null: false                    | 支払金額 |
| card_number    | integer    | null: false                    | カード情報 |
| expiration_date| integer    | null: false                    | 有効期限 |
| security_code  | integer    | null: false                    | セキュリティコード |
```
これらのカラムは不要ですので削除しましょう！

minachanchan commented on README.md:
```diff
@@ -1,24 +1,52 @@
+|-------------------|------------|--------------------------------|------|
+| purchase_record   | references | null: false, foreign_key: true | 購入履歴テーブルのID、purchase_recordsテーブルの外部キー |
+| postal_code       | string     | null: false                    | 郵便番号 |
+| prefecture_id     | integer    | null: false                    | 都道府県(アクティブハッシュ) |
```
```ruby
| prefecture_id | integer | null: false | 都道府県(アクティブハッシュ) |
```
都道府県を保存するカラムでは、商品情報テーブルで使用したactive_hashのデータを兼用できます。
商品情報で指定したものと同じものをこちらでも指定しましょう！

minachanchan commented on README.md:
```diff
@@ -1,24 +1,52 @@
+| prefecture_id     | integer    | null: false                    | 都道府県(アクティブハッシュ) |
+| city              | string     | null: false                    | 市区町村 |
+| address_number    | string     | null: false                    | 番地 |
+| building_name     | string     | null: false                    | 建物名 |
```
```ruby
| building_name | string | null: false | 建物名 |
```
建物名の入力は任意のため、null: falseは不要でございます。

minachanchan reviewed (commented):
実装お疲れ様です！レビューのご依頼ありがとうございます。
着実に完成に近づいておりますね！

コメントいたしましたので、ご確認頂けますと幸いでございます。

指摘内容の修正がお済みになりましたら、再度レビュー依頼くださいませ。
その際、ER図もREADMEの内容と合わせて修正とお願いいたします。
**次回のレビュー依頼の際に最新のER図のURLをまたこちらのプルリクエストに添付頂ければと存じます。**

実装完了まで引き続き一緒に頑張りましょう！
何卒よろしくお願いいたします。

Yokko405 committed:
データベース設計PR修正2

1.アソシエーションの追加：READMEにアソシエーションを追加。

2.カラムの削除: order_date、payment_amount、card_number、expiration_date、security_codeのカラムをpurchase_recordsテーブルから削除。

3.カラム名の兼用: itemsテーブルのprefecture_idカラムをshipping_addressesテーブルで兼用。

4.任意入力のカラム変更: building_nameカラムのnull: false制約を削除。

Yokko405 commented:
ご指摘いただきました4点について、修正しましたので、ご確認をお願いします。

▼ER図
https://gyazo.com/47d587e87d5ec8bb73f3084a606a0fec

minachanchan commented on README.md:
```diff
@@ -1,24 +1,69 @@
+| payment_amount | integer    | null: false                    | 支払金額 |
+| card_number    | integer    | null: false                    | カード情報 |
+| expiration_date| integer    | null: false                    | 有効期限 |
+| security_code  | integer    | null: false                    | セキュリティコード |
```
```ruby
| order_date     | date       | null: false                    | 注文日 |
| payment_amount | integer    | null: false                    | 支払金額 |
| card_number    | integer    | null: false                    | カード情報 |
| expiration_date| integer    | null: false                    | 有効期限 |
| security_code  | integer    | null: false                    | セキュリティコード |
```
[こちら](https://github.com/Yokko405/furima-40407/pull/12#discussion_r1524523183)についてご確認をお願いいたします！

minachanchan commented on README.md:
```diff
@@ -1,24 +1,69 @@
+|-------------------|------------|--------------------------------|------|
+| purchase_record   | references | null: false, foreign_key: true | 購入履歴テーブルのID、purchase_recordsテーブルの外部キー |
+| postal_code       | string     | null: false                    | 郵便番号 |
+| prefecture_id     | integer    | null: false                    | 都道府県(アクティブハッシュ) |
```
```ruby
| prefecture_id | integer | null: false | 都道府県(アクティブハッシュ) |
```
[こちら](https://github.com/Yokko405/furima-40407/pull/12#discussion_r1524524961)についてご確認をお願いいたします！

minachanchan reviewed (commented):
実装お疲れ様です！レビューのご依頼ありがとうございます。
着実に完成に近づいておりますね！

コメントいたしましたので、ご確認頂けますと幸いでございます。

指摘内容の修正がお済みになりましたら、再度レビュー依頼くださいませ。
その際、ER図もREADMEの内容と合わせて修正とお願いいたします。
**次回のレビュー依頼の際に最新のER図のURLをまたこちらのプルリクエストに添付頂ければと存じます。**

実装完了まであと少し、一緒に頑張りましょう！
何卒よろしくお願いいたします。

Yokko405 committed:
データベース設計PR修正3

・itemsテーブルのshipping_origin_idをprefecture_idにカラム名修正
・purchase_recordsテーブルの不要カラムを削除

Yokko405 commented:
すみません。READMEへの反映が漏れていました。
修正しましたので、ご確認をお願いします。

▼ER図
https://gyazo.com/47d587e87d5ec8bb73f3084a606a0fec

minachanchan commented on README.md:
```diff
@@ -1,24 +1,64 @@
+| condition_id            | integer | null: false               | 商品の状態(アクティブハッシュ) |
+| shipping_fee_payer_id   | integer | null: false               | 配送料の負担(アクティブハッシュ) |
+| prefecture_id           | integer | null: false               | 配送元の地域(アクティブハッシュ) |
+| shipping_days_id        | integer | null: false               | 発送までの日数(アクティブハッシュ) |
```
```ruby
| shipping_days_id | integer | null: false | 発送までの日数(アクティブハッシュ) |
```
指摘が重なり申し訳ございません。
active_hashで実装するカラムは「単数形_id」の形にする必要がございます。
「shipping_day_id」に変更をお願いいたします！

minachanchan reviewed (commented):
実装お疲れ様です！レビューのご依頼ありがとうございます。
キャプチャのご提示もお礼申し上げます。

コメントいたしましたので、ご確認頂けますと幸いでございます。

指摘内容の修正がお済みになりましたら、再度レビュー依頼くださいませ。
その際、ER図もREADMEの内容と合わせて修正とお願いいたします。
**次回のレビュー依頼の際に最新のER図のURLをまたこちらのプルリクエストに添付頂ければと存じます。**


何卒よろしくお願いいたします。

Yokko405 committed:
データベース設計PU修正4

itemsテーブルのshipping_day_idカラム修正

Yokko405 commented:
READMEとER図のitemsテーブルのカラム修正が終わりました。
→「shipping_day_id」に変更
ご確認をお願いします。

▼ER図
https://gyazo.com/3022944679747a6979425c00b3bef558

div-reviewer reviewed (commented):
Yokko405様、学習お疲れさまです。
レビュー依頼ありがとうございます！

前回のコメントにあった修正点が綺麗に直せましたね。素晴らしいです。
こちら`LGTM`とさせていただきます！実装お疲れさまでした。

【こちらは次回の機能に入る際の注意点です】
```ruby
プルリクエストのMerge、ローカルのmainブランチプルを忘れずに行い、ブランチを切ってから次の実装に取り組んで頂ければと思います。
```

**ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください！
次は「ユーザー管理機能」かと思われます！引き続き頑張りましょう！応援しております！**
